File: 2020/23.cpp
#include <bits/stdc++.h>

using namespace std;

#define INF numeric_limits<ll>::max()

typedef long long ll;
typedef vector<ll> vl;
typedef vector<char> vc;
typedef pair<ll, ll> pll;


vl load_input() {
    ifstream fin("23.txt");
    char x;
    vl puzzle;
    while (fin >> x) {
        puzzle.push_back(x - '0');    
    }
    return puzzle;
}

ll move(vl &state, ll current) {
    ll first = state[current];
    ll second = state[first];
    ll third = state[second];
    ll suffix = state[third];

    ll dst = current;
    while (dst==current || dst==first || dst==second || dst==third) {
        if (!--dst) dst = state.size() - 1;
    }

    state[current] = suffix;
    state[third] = state[dst];
    state[dst] = first;

    return state[current];
}

ll solve1(vl nums) {
    vl state(10);
    for (ll i = 0; i < nums.size(); i++) {
        state[nums[i]] = nums[(i + 1) % nums.size()];
    }
    
    ll current = nums[0];
    for (ll i = 0; i < 100; i++) {
        current = move(state, current);
    }
    
    ll out = 0;
    ll x = 1;
    while ((x = state[x]) != 1) {
        out = out * 10 + x;
    }
    
    return out;
}

ll solve2(vl nums) {
    vl state(1000001);
    iota(state.begin(), state.end(), 1);
    
    state[1000000] = nums[0];
    for (ll i = 1; i < nums.size(); i++) {
        state[nums[i - 1]] = nums[i];
    }
    state[nums.back()] = nums.size() + 1;
    
    ll current = nums[0];
    for (ll i = 0; i < 10000000; i++) {
        current = move(state, current);
    }
    
    ll first = state[1];
    ll second = state[first];
    
    return first * second;
}

int main() {
    vl puzzle = load_input();
    
    cout << "Part 1: " << solve1(puzzle) << "\n";
    cout << "Part 2: " << solve2(puzzle) << "\n";
    
    return 0;
}

File: Python/2020/23.py
from lib import *

input = read_input(2020, 23)


def move1(state):
    current, moving, suffix = state[0], state[1:4], state[4:]
    dst = suffix.index(max([x for x in suffix if int(x) < int(current)], default=max(suffix)))
    state = current + suffix[: dst + 1] + moving + suffix[dst + 1 :]
    return state[1:] + state[0]


state = input.strip()
for _ in range(100):
    state = move1(state)
one = state.index("1")
print(state[one + 1 :] + state[:one])


def move2(state, current):
    first = state[current]
    second = state[first]
    third = state[second]
    suffix = state[third]
    dst = current
    while dst in (current, first, second, third):
        dst -= 1
        if not dst:
            dst = len(state) - 1

    state[current] = suffix
    state[third] = state[dst]
    state[dst] = first
    return state[current]


nums = list(map(int, input.strip()))
state = list(range(1, 1000002))
for a, b in zip([-1] + nums, nums + [len(nums) + 1]):
    state[a] = b

current = nums[0]
for _ in range(10000000):
    current = move2(state, current)
first = state[1]
second = state[first]
print(first * second)
